compiler/src/iree/compiler/GlobalOptimization/SetEncoding.cpp

```diff
@@ -249,8 +276,12 @@ struct SetBatchMatmulEncoding : public OpRewritePattern<linalg::BatchMatmulOp> {
       }
       return {};
     };
-    Type lhsElemType = getElemType(origLhs);
-    Type rhsElemType = getElemType(origRhs);
+    bool lhsCasted = (bool)getCastElemType(origLhs);
+    bool rhsCasted = (bool)getCastElemType(origRhs);
+    Type lhsElemType =
+        lhsCasted ? getCastElemType(origLhs) : getElemType(origLhs);
+    Type rhsElemType =
+        rhsCasted ? getCastElemType(origRhs) : getElemType(origRhs);
```

Comment:
I think it is better to capture `std::optional<CastOpInterface>` here, so you don't need to get the defining op again in other functions. So this can be 

```cpp
std::optional<CastOpInterface> maybeLhsCastOp = getDefiningCastOp(origLhs);
Type lhsElemType = maybeLhsCastOp ? getCastElemType(maybeLhsCastOp.value()) : getElemType(origLhs);
```

Then you can pass the optional CastOpInterface to `padAndSetEncoding`